src/realm/history.cpp

```diff
@@ -27,7 +27,197 @@ using namespace realm;
+                                               ref_type new_top_ref)
+{
+    using gf = _impl::GroupFriend;
+    gf::remap(*m_group, new_file_size); // Throws
+    Allocator& alloc = gf::get_alloc(*m_group);
+    ref_type hist_ref = gf::get_history_ref(alloc, new_top_ref);
+    update_from_ref(hist_ref, new_version); // Throws
+}
+
+
+void InRealmHistory::update_from_parent(version_type version)
+{
+    using gf = _impl::GroupFriend;
+    ref_type ref = gf::get_history_ref(*m_group);
+    update_from_ref(ref, version); // Throws
+}
+
+
+void InRealmHistory::get_changesets(version_type begin_version, version_type end_version,
+                                    BinaryIterator* buffer) const noexcept
+{
+    REALM_ASSERT(begin_version <= end_version);
+    REALM_ASSERT(begin_version >= m_base_version);
+    REALM_ASSERT(end_version <= m_base_version + m_size);
+    version_type n_version_type = end_version - begin_version;
+    version_type offset_version_type = begin_version - m_base_version;
+    REALM_ASSERT(!util::int_cast_has_overflow<size_t>(n_version_type) &&
+                 !util::int_cast_has_overflow<size_t>(offset_version_type));
+    size_t n = size_t(n_version_type);
+    size_t offset = size_t(offset_version_type);
+    for (size_t i = 0; i < n; ++i)
+        buffer[i] = BinaryIterator(m_changesets.get(), offset + i);
+}
+
+
+void InRealmHistory::set_oldest_bound_version(version_type version)
+{
+    REALM_ASSERT(version >= m_base_version);
+    if (version > m_base_version) {
+        REALM_ASSERT(m_changesets);
+        size_t num_entries_to_erase = size_t(version - m_base_version);
+        // The new changeset is always added before set_oldest_bound_version()
+        // is called. Therefore, the trimming operation can never leave the
+        // history empty.
+        REALM_ASSERT(num_entries_to_erase < m_size);
+        for (size_t i = 0; i < num_entries_to_erase; ++i)
+            m_changesets->erase(0); // Throws
+        m_base_version += num_entries_to_erase;
+        m_size -= num_entries_to_erase;
+    }
+}
+
+
+void InRealmHistory::verify() const
+{
+#ifdef REALM_DEBUG
+    if (m_changesets)
+        m_changesets->verify();
+#endif
+}
+
+
+void InRealmHistory::update_from_ref(ref_type ref, version_type version)
+{
+    using gf = _impl::GroupFriend;
+    if (ref == 0) {
+        // No history
+        m_base_version = version;
+        m_size = 0;
+        m_changesets = nullptr;
+    }
+    else {
+        if (REALM_LIKELY(m_changesets)) {
+            m_changesets->update_from_ref(ref); // Throws
+        }
+        else {
+            Allocator& alloc = gf::get_alloc(*m_group);
+            bool nullable = false;
+            m_changesets = std::make_unique<BinaryColumn>(alloc, ref, nullable); // Throws
+            gf::set_history_parent(*m_group, *m_changesets->get_root_array());
+        }
+        m_size = m_changesets->size();
+        m_base_version = version - m_size;
+    }
+}
```

Comment:
Note: The entire `InRealmHistory` class was moved from `<realm/impl/continuous_transactions_history.hpp>`. It never belonged there.